akka-projection-slick/src/main/scala/akka/projection/slick/SlickProjection.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.slick
+
+import akka.Done
+import akka.actor.ClassicActorSystemProvider
+import akka.annotation.{ ApiMayChange, InternalApi }
+import akka.projection.Projection
+import akka.stream.scaladsl.{ Keep, Sink, Source }
+import akka.stream.{ KillSwitch, KillSwitches }
+import org.slf4j.LoggerFactory
+import slick.basic.DatabaseConfig
+import slick.dbio.DBIO
+import slick.jdbc.JdbcProfile
+
+import scala.concurrent.Future
+import scala.reflect.ClassTag
+
+object SlickProjection {
+  @ApiMayChange
+  def transactional[Offset, StreamElement, P <: JdbcProfile: ClassTag](
+      projectionId: String,
+      sourceProvider: Option[Offset] => Source[StreamElement, _],
+      offsetExtractor: StreamElement => Offset,
+      databaseConfig: DatabaseConfig[P])(eventHandler: StreamElement => DBIO[Done]): Projection =
+    new SlickProjectionImpl(projectionId, sourceProvider, offsetExtractor, databaseConfig, eventHandler)
+}
+
+@InternalApi
+private[projection] class SlickProjectionImpl[Offset, StreamElement, P <: JdbcProfile](
+    projectionId: String,
+    sourceProvider: Option[Offset] => Source[StreamElement, _],
+    offsetExtractor: StreamElement => Offset,
+    databaseConfig: DatabaseConfig[P],
+    eventHandler: StreamElement => DBIO[Done])
+    extends Projection {
+
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  private val offsetStore = new OffsetStore(databaseConfig.db, databaseConfig.profile)
+
+  private var shutdown: Option[KillSwitch] = None
+
+  override def start()(implicit systemProvider: ClassicActorSystemProvider): Unit = {
+
+    implicit val dispatcher = systemProvider.classicSystem.dispatcher
+    import databaseConfig.profile.api._
+
+    val lastKnownOffset: Future[Option[Offset]] = offsetStore.readOffset(projectionId)
+
+    val killSwitch =
+      Source
+        .future(lastKnownOffset.map { offsetOpt =>
+          logger.debug(s"Starting projection '$projectionId' from offset '$offsetOpt' ")
+          sourceProvider(offsetOpt)
+        })
+        .flatMapConcat[StreamElement, Any](identity)
+        .mapAsync(1) { elt =>
+          // run user function and offset storage on the same transaction
+          // offset storage comes last, any side-effect in user function is therefore at-least-once
+          val txDBIO =
+            eventHandler(elt).flatMap(_ => offsetStore.saveOffset(projectionId, offsetExtractor(elt))).transactionally
```

Comment:
Yep, at-least-once for side-effects. Not sure what is best here. The reality is that this is that side-effect inside a DB transaction is a bad thing anyway. 

I don't think we should make a big deal of it. Otherwise, we need to either provide a `transactional` with `at-least-once` side-effect and a `transactional` with `at-most-once` side-effect. Not really excited about such a thing. Maybe better to simply document that side-effects will be run `at-least-once`.

Reply:
You mean the order of eventHandler and saveOffset?

I agree that the primary side effect is the returned DBIO.

If the only side effect is the DBIO it will always be exactly-once, independent of the order of eventHandler and saveOffset.

If there is some extra external side effect in the eventHandler in addition to the returned DBIO that side effect will be at-least-once. It will be at-least-once independent of the order because full transaction is rolled back if any of it fails.

However, we can make it a bit more best effort also for such additional side effect if we have saveOffset first. Assuming that the most common type of problem will be connection error that will be detected first by the saveOffset and thereby the eventHandler is not called at all. Still at-least-once, but with less chance of running additional side effect more than once.

Also, is there something like a DBIO.onSuccess(_ => ) that is called when transaction is committed? Then such extra external side effect can be placed in such callback if at-most-once is needed.

Reply:
I was thinking of side-effect other than the DB operation. The database IO is the effect we want, not a side-effect. 

About the ordering, there are a few scenarios:

If the user writes:

```scala
def eventHandler(evt:Event): DBIO = {
   runSomeNastySideEffects()
   DBIO(...) // <- real dbio stuff
}
```

If we chain it as `eventHandler andThen offset`, the side-effect will run before we even try to use a connection. So, yes. Better to have the offset first indeed.

But in Slick, you can also add a side-effect as such:

```scala
def eventHandler(evt:Event): DBIO = {
   for { 
	 _ <- DBIO.successful(runSomeNastySideEffects()) 
     _ <- DBIO(...) // <- real dbio stuff
   } yield Done
}
```

In such a case, the side-effect will run inside the transactional block thus after we get a connection. If we fail to get a connection, it won't run. It will still be `at-least-once`.
However, if get a connection we can execute part of the IOs and then fail when closing the connection or committing the transaction. The effect for the user is the same and order of chaining won't matter anymore.

Reply:
I revert the order and added a few other tests:

* fail using DBIO.failed
* fail while throwing an exception
* fail on user code while inserting illegal null value